## 1. Craps
This is a game played with two dice. You decibe to play. Each bet is one chip. The goal is to roll a seven (with two dice). The house pays 4 chips plus your original chip if you win. Is this fair? (If not, define fair).

First let's look at probability we win

There are 36 possible outcomes from rolling 2 dice (6 X 6 = 36)

There are 6 combinations that can result in 7 (1,6), (2,5), (3,4), (4,3), (5,2), (6,1)

Therefore odd of rolling a 7 with two dice is 6/36 = 1/6

The probability of not winning is therefore 5/6

To determine the expected pay we can use the formula:

**E = (P(win) X payout for win) + (P(lose) X payout for lose)**

In [4]:
E = ((1/6)*5)+((5/6)*-1)
print(abs(round(E, 2)))

0.0


Since the expected payout is 0 you neither gain nor lose on average. This also suggest that the game is **fair** since the house does not have a mathematical advantage on you.

## 9. Rolling three sums that are not prime

You roll three dice. What is the probability that the sum of the three numbers is not a prime number? How can you verify this answer?

In [7]:
from itertools import product
from sympy import isprime

# Generate all possible sums of three dice rolls
sums = [sum(roll) for roll in product(range(1, 7), repeat=3)]

# Identify prime sums
prime_sums = [s for s in sums if isprime(s)]

# Calculate the probability
total_rolls = len(sums)
non_prime_rolls = total_rolls - len(prime_sums)
probability_non_prime = non_prime_rolls / total_rolls

print(f"Probability that the sum is not a prime number: {probability_non_prime:.4f}")

[3, 4, 5, 6, 7, 8, 4, 5, 6, 7, 8, 9, 5, 6, 7, 8, 9, 10, 6, 7, 8, 9, 10, 11, 7, 8, 9, 10, 11, 12, 8, 9, 10, 11, 12, 13, 4, 5, 6, 7, 8, 9, 5, 6, 7, 8, 9, 10, 6, 7, 8, 9, 10, 11, 7, 8, 9, 10, 11, 12, 8, 9, 10, 11, 12, 13, 9, 10, 11, 12, 13, 14, 5, 6, 7, 8, 9, 10, 6, 7, 8, 9, 10, 11, 7, 8, 9, 10, 11, 12, 8, 9, 10, 11, 12, 13, 9, 10, 11, 12, 13, 14, 10, 11, 12, 13, 14, 15, 6, 7, 8, 9, 10, 11, 7, 8, 9, 10, 11, 12, 8, 9, 10, 11, 12, 13, 9, 10, 11, 12, 13, 14, 10, 11, 12, 13, 14, 15, 11, 12, 13, 14, 15, 16, 7, 8, 9, 10, 11, 12, 8, 9, 10, 11, 12, 13, 9, 10, 11, 12, 13, 14, 10, 11, 12, 13, 14, 15, 11, 12, 13, 14, 15, 16, 12, 13, 14, 15, 16, 17, 8, 9, 10, 11, 12, 13, 9, 10, 11, 12, 13, 14, 10, 11, 12, 13, 14, 15, 11, 12, 13, 14, 15, 16, 12, 13, 14, 15, 16, 17, 13, 14, 15, 16, 17, 18]
Probability that the sum is not a prime number: 0.6620


As seen from the above calculated result, the probabiliyt of getting a non-prime number is **66.2%**

## Locker Problem
A school’s lockers are numbered 1 to 100. One hundred students
enter the school one at a time. The first student opens the lockers.
The second student closes the even numbered lockers. The third
student either closes or opens every third locker. The remaining
students continue the pattern. Write a program to determine
which lockers remain open at the end, after the last one of the
students passes. Explain the result. Make sure the program not
only reports the situation at the end but also conveys the stages
through which the array of lockers passes.

In [14]:
def toggle_lockers():
    # Initialize lockers: False means closed, True means open
    lockers = [False] * 100

    # Iterate through each student
    for student in range(1, 101):
        for locker in range(student - 1, 100, student):
            lockers[locker] = not lockers[locker]

    # Determine which lockers remain open
    open_lockers = [i + 1 for i, locker in enumerate(lockers) if locker]
    print(f"Lockers that remain open: {open_lockers}")

toggle_lockers()

Lockers that remain open: [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


As we can see from the results, **the lockers that remain open are perfect squares from 1 – 100**

## Coin Flip Streaks

Write a program to find out how often a streak of six heads or a streak
of six tails comes up in a randomly generated list of heads and tails. Your
program breaks up the experiment into two parts: the first part generates
a list of randomly selected 'heads' and 'tails' values, and the second part
checks if there is a streak in it. Put all of this code in a loop that repeats the experiment 10,000 times so we can find out what percentage of the coin flips contains a streak of six heads or tails in a row. As a hint, the function call random.randint(0, 1) will return a 0 value 50% of the time and a 1 value the other 50% of the time.

In [16]:
import random

def generate_coin_flips(num_flips):
    coin_flips = []
    for _ in range(num_flips):
        flip = random.randint(0, 1)
        if flip == 0:
            coin_flips.append('heads')
        else:
            coin_flips.append('tails')
    return coin_flips

num_experiments = 10000
streak_length = 6
streak_count = 0

for _ in range(num_experiments):
    coin_flips = generate_coin_flips(100)  
    current_streak = 1
    
    for i in range(1, len(coin_flips)):
        if coin_flips[i] == coin_flips[i - 1]:
            current_streak += 1
        else:
            current_streak = 1
        
        if current_streak == streak_length:
            streak_count += 1
            break  
percentage = (streak_count / num_experiments) * 100
print(f"Percentage of coin flips with a streak of {streak_length} heads or tails: {percentage:.2f}%")


Percentage of coin flips with a streak of 6 heads or tails: 80.45%


**The percenatge of getting 6 heads or tails in a row is 80.45%**

## 1. First Problem

In [17]:
class Complex:
    def __init__(self, real, imag):
        self.real = real
        self.imag = imag

    def __add__(self, other):
        return Complex(self.real + other.real, self.imag + other.imag)

    def __mul__(self, other):
        real_part = self.real * other.real - self.imag * other.imag
        imag_part = self.real * other.imag + self.imag * other.real
        return Complex(real_part, imag_part)

    def __str__(self):
        return f"{self.real} + {self.imag}i"
    
c1 = Complex(1, 2)
c2 = Complex(2, 3)

print(c1 + c2)

print(c1 * c2)

3 + 5i
-4 + 7i


## 2. Second Problem

In [18]:
class Creature:
    def __init__(self, name):
        self.name = name
    
    def speak(self):
        print(f"My name is {self.name} and I am a Creature.")

class Dog(Creature):
    def speak(self):
        print(f"My name is {self.name} and I am a Dog: woof.")

class Cow(Creature):
    def speak(self):
        print(f"My name is {self.name} and I am a Cow: moo.")

creature = Creature("Bob")
dog = Dog("Fido")
cow = Cow("Betsy")

creature.speak()
dog.speak()
cow.speak()

My name is Bob and I am a Creature.
My name is Fido and I am a Dog: woof.
My name is Betsy and I am a Cow: moo.
